File: php/src/Shinchoku.php
<?php

declare(strict_types=1);

namespace Shinchoku;

/**
 * shinchoku (進捗, "progress") — a tiny NDJSON protocol for batch CLIs to
 * report their progress. One JSON object per line on stdout.
 *
 * The contract is the spec (spec/SPEC.md in the repository), not this class:
 * `echo '{"v":1,"event":"done"}' . "\n";` is fully compliant. This class is
 * the convenience helper for PHP producers.
 */
final class Shinchoku
{
    /** Protocol version stamped into every line as "v". */
    public const VERSION = 1;

    /** @var resource|null Output stream; defaults to STDOUT. */
    private static $out = null;

    private function __construct()
    {
    }

    /**
     * Redirect event output (default: STDOUT). Intended for tests and for
     * consumers that relay events over another transport.
     *
     * @param resource $stream
     */
    public static function setOutput($stream): void
    {
        self::$out = $stream;
    }

    /** Announce the run. Should be the first event. */
    public static function start(?string $title = null, ?int $total = null): void
    {
        self::emit(['event' => 'start', 'title' => $title, 'total' => $total]);
    }

    /** Report `current` completed units. Null total means unknown. */
    public static function progress(int $current, ?int $total = null, ?string $msg = null): void
    {
        self::emit(['event' => 'progress', 'current' => $current, 'total' => $total, 'msg' => $msg]);
    }

    /** Emit a log line with a severity ("info" | "warn" | "error"). */
    public static function log(string $level, string $msg): void
    {
        self::emit(['event' => 'log', 'level' => $level, 'msg' => $msg]);
    }

    /** log("info", $msg) */
    public static function info(string $msg): void
    {
        self::log('info', $msg);
    }

    /** log("warn", $msg) */
    public static function warn(string $msg): void
    {
        self::log('warn', $msg);
    }

    /** log("error", $msg) — non-fatal; a fatal error is failed(). */
    public static function error(string $msg): void
    {
        self::log('error', $msg);
    }

    /**
     * Emit a named value (number or string) for the consumer to display
     * verbatim. The key is data, not vocabulary.
     */
    public static function metric(string $key, int|float|string $value): void
    {
        self::emit(['event' => 'metric', 'key' => $key, 'value' => $value]);
    }

    /** Declare that an output file exists. */
    public static function artifact(string $path, ?string $label = null): void
    {
        self::emit(['event' => 'artifact', 'path' => $path, 'label' => $label]);
    }

    /** Report successful completion. */
    public static function done(?string $summary = null): void
    {
        self::emit(['event' => 'done', 'summary' => $summary]);
    }

    /** Report fatal failure; the run is over. */
    public static function failed(string $msg): void
    {
        self::emit(['event' => 'failed', 'msg' => $msg]);
    }

    /**
     * Emit one NDJSON line with "v" first, dropping null fields. Best-effort:
     * write errors are ignored, matching the fire-and-forget nature of reporting.
     *
     * @param array<string, mixed> $fields
     */
    private static function emit(array $fields): void
    {
        $fields = array_filter($fields, static fn (mixed $value): bool => $value !== null);
        $line = json_encode(
            ['v' => self::VERSION] + $fields,
            JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES
        );
        if ($line === false) {
            return;
        }
        @fwrite(self::$out ?? STDOUT, $line . "\n");
    }
}

File: go/shinchoku.go
// Package shinchoku emits shinchoku (進捗, "progress") protocol events:
// a tiny NDJSON protocol for batch CLIs to report their progress on stdout.
//
// The contract is the spec (spec/SPEC.md in the repository), not this package:
// printing `{"v":1,"event":"done"}` yourself is fully compliant. This package
// is the convenience helper for Go producers.
//
//	shinchoku.Start("import", 120)
//	shinchoku.Progress(3, 120)
//	shinchoku.Done("4,520 items")
package shinchoku

import (
	"encoding/json"
	"io"
	"os"
	"sync"
)

// Version is the protocol version stamped into every line as "v".
const Version = 1

// Level is the severity of a Log event.
type Level string

const (
	LevelInfo  Level = "info"
	LevelWarn  Level = "warn"
	LevelError Level = "error"
)

var (
	mu  sync.Mutex
	out io.Writer = os.Stdout
)

// SetOutput redirects event output (default: os.Stdout). Intended for tests
// and for consumers that relay events over another transport.
func SetOutput(w io.Writer) {
	mu.Lock()
	defer mu.Unlock()
	out = w
}

// emit marshals fields (plus "v") as one NDJSON line, best-effort: write
// errors are ignored, matching the fire-and-forget nature of reporting.
// json.Marshal on a map sorts keys, so output is deterministic.
func emit(fields map[string]any) {
	fields["v"] = Version
	line, err := json.Marshal(fields)
	if err != nil {
		return
	}
	mu.Lock()
	defer mu.Unlock()
	_, _ = out.Write(append(line, '\n'))
}

// Start announces the run. Empty title / zero total are omitted.
func Start(title string, total uint64) {
	f := map[string]any{"event": "start"}
	if title != "" {
		f["title"] = title
	}
	if total > 0 {
		f["total"] = total
	}
	emit(f)
}

// Progress reports current completed units. Zero total means unknown.
func Progress(current, total uint64) {
	ProgressMsg(current, total, "")
}

// ProgressMsg is Progress with a short human-readable detail (URL, filename…).
func ProgressMsg(current, total uint64, msg string) {
	f := map[string]any{"event": "progress", "current": current}
	if total > 0 {
		f["total"] = total
	}
	if msg != "" {
		f["msg"] = msg
	}
	emit(f)
}

// Log emits a log line with a severity.
func Log(level Level, msg string) {
	emit(map[string]any{"event": "log", "level": string(level), "msg": msg})
}

// Info is Log(LevelInfo, msg).
func Info(msg string) { Log(LevelInfo, msg) }

// Warn is Log(LevelWarn, msg).
func Warn(msg string) { Log(LevelWarn, msg) }

// Error is Log(LevelError, msg) — a non-fatal error; a fatal one is Failed.
func Error(msg string) { Log(LevelError, msg) }

// Metric emits a named value (number or string) for the consumer to display
// verbatim. The key is data, not vocabulary.
func Metric(key string, value any) {
	emit(map[string]any{"event": "metric", "key": key, "value": value})
}

// Artifact declares that an output file exists. Empty label is omitted.
func Artifact(path, label string) {
	f := map[string]any{"event": "artifact", "path": path}
	if label != "" {
		f["label"] = label
	}
	emit(f)
}

// Done reports successful completion. Empty summary is omitted.
func Done(summary string) {
	f := map[string]any{"event": "done"}
	if summary != "" {
		f["summary"] = summary
	}
	emit(f)
}

// Failed reports fatal failure; the run is over.
func Failed(msg string) {
	emit(map[string]any{"event": "failed", "msg": msg})
}
